modules/cli/src/integrationTest/java/org/apache/ignite/cli/ItClusterCommandTest.java

```diff
@@ -0,0 +1,256 @@
+        String config = Files.readString(Path.of("src/integrationTest/resources/hardcoded-ports-config.json"));
+        config = config.replaceAll("<NETWORK_PORT>", String.valueOf(node.networkPort));
+        config = config.replaceAll("<REST_PORT>", String.valueOf(node.restPort));
+        config = config.replaceAll("<NET_CLUSTER_NODES>", netClusterNodes());
+
+        return config;
+    }
+
+    private String netClusterNodes() {
+        return NODES.stream()
+                .map(Node::networkHostPort)
+                .map(s -> "\"" + s + "\"")
+                .collect(joining(", ", "[", "]"));
+    }
+
+    private boolean isHttpServedAt(String hostPort) {
+        HttpRequest request = HttpRequest.newBuilder(uriFor(hostPort))
+                .GET()
+                .build();
+
+        try {
+            HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.discarding());
+
+            return true;
+        } catch (IOException e) {
+            return false;
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return false;
+        }
+    }
+
+    private URI uriFor(String hostPort) {
+        try {
+            return new URI("http://" + hostPort + "/");
+        } catch (URISyntaxException e) {
+            throw new RuntimeException("Cannot parse URI", e);
+        }
+    }
+
+    @AfterEach
+    void tearDown(TestInfo testInfo) {
+        for (int i = 0; i < 3; i++) {
+            IgnitionManager.stop(testNodeName(testInfo, i));
+        }
+
+        if (ctx != null) {
+            ctx.stop();
+        }
+    }
+
+    /**
+     * Starts a cluster of 4 nodes and executes init command on it. First node is used to issue the command via REST endpoint,
+     * second will host the meta-storage RAFT group, third will host the Cluster Management RAFT Group (CMG), fourth
+     * will be just a node.
+     *
+     * @param testInfo test info (used to derive node names)
+     */
+    @Test
+    void initClusterWithNodesOfDifferentRoles(TestInfo testInfo) {
+        int exitCode = cmd(ctx).execute(
+                "cluster", "init",
+                "--node-endpoint", FIRST_NODE.restHostPort(),
+                "--meta-storage-node", SECOND_NODE.nodeName(testInfo),
+                "--cmg-node", THIRD_NODE.nodeName(testInfo)
+        );
+
+        assertThat(
+                String.format("Wrong exit code; std is '%s', stderr is '%s'", out.toString(UTF_8), err.toString(UTF_8)),
+                exitCode, is(0)
+        );
+        assertThat(out.toString(UTF_8), is("Cluster was initialized successfully." + NL));
+
+        // TODO: when IGNITE-16526 is implemented, also check that the logical topology contains all 4 nodes
+    }
+
+    /**
+     * Creates a new command line interpreter.
+     *
+     * @param applicationCtx DI context.
+     * @return New command line instance.
+     */
+    private CommandLine cmd(ApplicationContext applicationCtx) {
```

Comment:
I think this method should be moved to `AbstractCliTest`

Reply:
This also requires to move `out` and `err` (or change something else), so, to reduce invasivity, I added another abstract superclass (`AbstractCliIntegrationTest`) and moved the method and two fields there.